[#9420] Consolidate Course RESTFul API

ChooJeremy opened the pull request:
Part of #9420 
<!-- Remove this portion after you have made the checks. -->

**Outline of Solution**

<!-- Tell us how you solved the issue. -->
<!-- If there are things you want the reviewers to focus on, include them here as well. -->
<!-- This portion can be skipped if the fix is trivial. -->
Other than just renaming, the following changes were made:
* Moved CourseInfo to output folder so it gets processed by generateTypes, extracted out into CourseData and CoursesData.
* Added CourseBasicRequest and their corresponding requests. I reused CourseInfo in CourseBasicRequest instead of re-declaring those variables like what FeedbackSessionBasicRequest because this way if CourseInfo changes the others change as well. However, this does require that the API caller add an extra wrapper (storedData in this case) to the JSON sent.
* Converted Create, Archive and Save request to use the body to transfer data instead of querystring. While doing this, I found a bug: if you add 
`verifyHttpParameterFailure(Const.ParamsNames.COURSE_ID, session.getCourseId(), Const.ParamsNames.FEEDBACK_SESSION_NAME, session.getFeedbackSessionName());`
to SaveFeedbackSessionActionTest, line 46, it will fail, despite it being a valid request the user can send (Valid querystring but no response body).  The result is a NullPointerException when parsing the body, as opposed to the usual HTTP based exceptions. To fix this, I delved into Action's `getAndValidateRequestBody` and added a null check to it.
* Added course.service.ts to be used for accessing the Course API. Note that the create and update calls don't actually have use for the parameter querystring, but it's still there to be passed in an empty list.
* There are 3 ways to remove courses. Soft, Hard and Archive. calling DELETE /course is now a hard delete. calling PUT /course/archive is to archive, and calling PUT /bin/course is for soft delete.
* The original test cases for SaveCourseActionTest included sending an invalid timezone to test if it would report an error appropriately.  This isn't quite feasible now because in order to create an invalid timezone, the wrong string has to be passed to `CourseSaveRequest`, who's only constructor takes in a `CourseData`, who's only constructor takes in a `CourseAttributes`, which requires a timezone of type ZoneId, so it's impossible to provide an invalid timezone. Normally I would add an additional constructor to `CourseData` that takes in Strings to populate the fields, but to avoid adding methods in production just for testing, instead I converted `CourseData.timeZone` to be non-final and added a setter for it.

Ready for review.

xpdavid commented on src/main/java/teammates/ui/webapi/action/SaveCourseAction.java:
```diff
@@ -12,7 +12,7 @@
 /**
  * Action: Save edited course details.
```
Don't forget to change the javadoc

xpdavid commented on src/main/java/teammates/ui/webapi/output/CourseData.java:
```diff
@@ -0,0 +1,39 @@
+import teammates.common.datatransfer.attributes.CourseAttributes;
+
+/**
+ * The API output format of a list of {@link CourseAttributes}.
```
Not matched javadoc?

xpdavid commented on src/main/java/teammates/ui/webapi/output/CoursesData.java:
```diff
@@ -0,0 +1,21 @@
+import teammates.common.datatransfer.attributes.CourseAttributes;
+
+/**
+ * Response of a list of courses.
```
`The API output format of a list of {@link CourseAttributes}.`?

xpdavid commented on src/web/services/course.service.ts:
```diff
@@ -0,0 +1,30 @@
+import { HttpRequestService } from './http-request.service';
+
+/**
+ * Handles sessions related logic provision.
```
Doc not matched?

xpdavid commented on src/web/services/course.service.ts:
```diff
@@ -0,0 +1,30 @@
+@Injectable({
+  providedIn: 'root',
+})
+export class CourseService {
```
You can use `ng g s CourseService` so that Angular CLI will generate corresponding test file also. Instead of copy here.

xpdavid commented on src/main/java/teammates/ui/webapi/request/CourseBasicRequest.java:
```diff
@@ -0,0 +1,23 @@
+ */
+public class CourseBasicRequest extends BasicRequest {
+
+    private CourseData storedData;
```
> * Added CourseBasicRequest and their corresponding requests. I reused CourseInfo in CourseBasicRequest instead of re-declaring those variables like what FeedbackSessionBasicRequest because this way if CourseInfo changes the others change as well.

I don't think this is a good reason to "reuse" course data.
- The wrapper gives extra confusion to other developers and does not follow the convention of other APIs.
- It is just in this case that the request and response have the same fields.

xpdavid commented on src/main/java/teammates/ui/webapi/action/DeleteCourseAction.java:
```diff
@@ -1,17 +1,15 @@
  */
 public class DeleteCourseAction extends Action {
 
+    private String idOfCourseToDelete;
```
Do you really need to do this?

ChooJeremy replied:
Um this was just a rename from DeleteInstructorSoftDeletedCourseAction > DeleteCourseAction. DeleteCourseAction used to be soft delete hence the changes. 
...
but yes it appears unnecessary. I'll remove it. IntelliiJ reports similar code in `RestoreInstructorSoftDeleteCourseAction` so that probably could be deleted as well.

xpdavid commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -930,6 +930,8 @@ private Const() {
         public static final String ACCOUNTS_DOWNGRADE = "/accounts/downgrade";
         public static final String COURSE = "/course";
+        public static final String COURSE_ARCHIVE = "/course/archive";
+        public static final String COURSE_SOFTDELETE = "/bin/course";
```
Follow the convention `BIN_COURSE = "/bin/course"`?

xpdavid commented on src/main/java/teammates/ui/webapi/action/ArchiveCourseAction.java:
```diff
@@ -25,8 +26,10 @@ public void checkSpecificAccessControl() {
-        String archiveStatus = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ARCHIVE_STATUS);
+        CourseArchiveRequest courseArchiveRequest = getAndValidateRequestBody(CourseArchiveRequest.class);
+
+        String idOfCourseToArchive = courseArchiveRequest.getCourseId();
```
The courseId is used to identify the resource. So it should not be inside request body.

xpdavid commented on src/main/java/teammates/ui/webapi/action/SaveCourseAction.java:
```diff
@@ -33,9 +34,11 @@ public void checkSpecificAccessControl() {
-        String courseTimeZone = getNonNullRequestParamValue(Const.ParamsNames.COURSE_TIME_ZONE);
+        CourseSaveRequest courseSaveRequest = getAndValidateRequestBody(CourseSaveRequest.class);
+
+        String courseId = courseSaveRequest.getCourseData().getCourseId();
```
Same here. CourseId is used to identify the resource to save.

xpdavid commented on src/main/java/teammates/ui/webapi/request/BasicRequest.java:
```diff
@@ -3,7 +3,11 @@
 
 /**
- * The request body of a HTTP request.
+ * A BasicRequest represents the request body of a HTTP request. Within the HTTP request, it is expected that the
```
It looks like you are talking about `how` to use it?

And also please avoid talking things in frontend like `web/types/api-output.ts`

ChooJeremy replied:
Yes. Kind of explaining how it is utilized. Because when I was looking through it it was very confusing to pierce everything together. 

How should I refer to it then? Just api-output.ts?

wkurniawan07 replied:
Do not make any reference to it at all. The web API stands on its own, with or without `api-output.ts`.

xpdavid commented on src/main/java/teammates/ui/webapi/request/CourseArchiveRequest.java:
```diff
@@ -0,0 +1,32 @@
+package teammates.ui.webapi.request;
+
+/**
+ * The delete request to archive a course.
```
Not matched javadoc.

xpdavid commented on src/web/services/course.service.ts:
```diff
@@ -0,0 +1,45 @@
+  /**
+   * Updates a course by calling API.
+   */
+  updateCourse(paramMap: { [key: string]: string }, request: CourseSaveRequest): Observable<MessageOutput> {
```
You should have something like `updateCourse(courseId: string, request: CourseSaveRequest)`?

Related:

> * Note that the create and update calls don't actually have use for the parameter querystring, but it's still there to be passed in an empty list.

Just remove the parameter.

ChooJeremy commented on src/main/java/teammates/ui/webapi/action/CreateCourseAction.java:
```diff
@@ -10,7 +10,7 @@
 /**
  * Action: Adds a new course for instructor.
```
Sorry - I'm not sure what's wrong with the javadoc. Looking at the samples CreateFeedbackQuestionAction, CreateFeedbackResponseAction etc their format is just "Create a feedback question" or "Create a feedback response". Do I just change this to "Create a new course for an instructor?"

xpdavid replied:
Yup since your action name is `create`.

ChooJeremy commented:
Ready for review.

Sorry about the extra parameters still being included in course.service.ts when you last reviewed. I kind of thought that since this was "just renaming" I'd just write up a wrapper function for the original calls and leave it as that.

Changes:
* Checked and edited javadocs for all files changed. Updated based on samples in feedback-session.
* Course ID is now moved to param in Create, Save, Archive and Delete. The rest are in body. The data in Request have been streamlined and reduced to the bare essentials.

xpdavid commented on src/main/java/teammates/ui/webapi/action/ActionFactory.java:
```diff
@@ -98,8 +99,7 @@
         map(ResourceURIs.COURSE_EDIT_DETAILS, GET, GetCourseEditDetailsAction.class);
-        map(ResourceURIs.COURSE_EDIT_DETAILS_SAVE, PUT, SaveCourseEditDetailsAction.class);
-        map(ResourceURIs.COURSE_DELETE, DELETE, DeleteCourseAction.class);
+        map(ResourceURIs.COURSE_DELETE, DELETE, SoftDeleteCourseAction.class);
```
Why there is a duplicated entry above `map(ResourceURIs.BIN_COURSE, PUT, SoftDeleteCourseAction.class);`?

xpdavid commented on src/main/java/teammates/ui/webapi/action/ArchiveCourseAction.java:
```diff
@@ -26,7 +27,10 @@ public void checkSpecificAccessControl() {
         String idOfCourseToArchive = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
-        String archiveStatus = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ARCHIVE_STATUS);
+
+        CourseArchiveRequest courseArchiveRequest = getAndValidateRequestBody(CourseArchiveRequest.class);
```
`CourseArchiveRequest` -> `ArchiveCourseRequest`?

ChooJeremy replied:
Sure, but this seems to break convention? All other requests are named <Object> <Action> <Request> i.e .CourseBasicRequest, CourseSaveRequest etc.

xpdavid replied:
Yup. Notice that. Sorry.

xpdavid commented on src/main/java/teammates/ui/webapi/action/CreateCourseAction.java:
```diff
@@ -27,8 +28,11 @@ public void checkSpecificAccessControl() {
     @Override
     public ActionResult execute() {
         String newCourseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
```
`COURSE_ID` is not something to identify the resource and therefore it should inside `CourseCreateRequest`

xpdavid commented on src/main/java/teammates/ui/webapi/action/CreateCourseAction.java:
```diff
@@ -27,8 +28,11 @@ public void checkSpecificAccessControl() {
-        String newCourseName = getNonNullRequestParamValue(Const.ParamsNames.COURSE_NAME);
-        String newCourseTimeZone = getNonNullRequestParamValue(Const.ParamsNames.COURSE_TIME_ZONE);
+
+        CourseCreateRequest courseCreateRequest = getAndValidateRequestBody(CourseCreateRequest.class);
```
`CourseCreateRequest ` -> `CreateCourseRequest`?

xpdavid commented on src/main/java/teammates/ui/webapi/request/CourseArchiveRequest.java:
```diff
@@ -0,0 +1,22 @@
+ * The archive request of a course.
+ */
+public class CourseArchiveRequest extends BasicRequest {
+    private String archiveStatus;
```
This can be `boolean`?

xpdavid commented on src/test/java/teammates/test/cases/webapi/ArchiveCourseActionTest.java:
```diff
@@ -28,71 +29,59 @@ protected String getRequestMethod() {
+        ArchiveCourseAction archiveAction = getAction(courseArchiveRequest, submissionParams);
         JsonResult result = getJsonResult(archiveAction);
 
         assertEquals(HttpStatus.SC_OK, result.getStatusCode());
```
Not your fault. I think we should add a verification (i.e. the course is indeed archived) here.

There are multiple places where the verification lacks. If you have time, please help to fix it.

xpdavid commented on src/test/java/teammates/test/cases/webapi/DeleteCourseActionTest.java:
```diff
@@ -30,87 +28,40 @@ protected String getRequestMethod() {
-        verifyHttpParameterFailure();
-
-        ______TS("Typical case, 2 courses");
+        ______TS("Typical case, delete a soft-deleted course in Recycle Bin");
```
Add a case where a course is not in recycle bin? -> it should fail

ChooJeremy replied:
It doesn't fail?

xpdavid replied:
If it just `execute()`, then it won't. But with access control `checkSpecificAccessControl`, it will. Try put the test in `testAccessControl`?

xpdavid commented on src/test/java/teammates/test/cases/webapi/DeleteCourseActionTest.java:
```diff
@@ -30,87 +28,40 @@ protected String getRequestMethod() {
-                Const.ParamsNames.COURSE_ID, "icdct.tpa.id1",
-        };
+        loginAsInstructor(instructorId);
+        CoursesLogic.inst().moveCourseToRecycleBin(courseId);
```
you can use `logic.moveCourseToRecycleBin(courseId);`

ChooJeremy commented:
Ready for review.

ChooJeremy commented:
Ready for review.

xpdavid commented on src/main/java/teammates/ui/webapi/action/BinCourseAction.java:
```diff
@@ -0,0 +1,43 @@
+import teammates.common.util.Const;
+
+/**
+ * Soft delete a course.
```
I prefer not to expose the term `soft-delete` in the application level. We should mention `Move a course to to recycle bin instead`.

xpdavid commented on src/main/java/teammates/ui/webapi/output/CourseData.java:
```diff
@@ -0,0 +1,43 @@
+        return timeZone;
+    }
+
+    public void setTimeZone(String s) {
```
Do you need setter?

ChooJeremy replied:
Ah yes, no. That was originally required because of BasicCourseAction being coupled to CourseData, and one of the tests using an invalid Time Zone. Since the coupling has been removed timeZone can now be set to final.

xpdavid commented on src/test/java/teammates/test/cases/webapi/ArchiveCourseActionTest.java:
```diff
@@ -28,71 +30,67 @@ protected String getRequestMethod() {
 
         ______TS("Not enough parameters");
         verifyHttpParameterFailure();
-        verifyHttpParameterFailure(Const.ParamsNames.COURSE_ID, courseId);
```
Do you have test case somewhere where parameters are there but request body is null?

xpdavid commented on src/test/java/teammates/test/cases/webapi/ArchiveCourseActionTest.java:
```diff
@@ -28,71 +30,67 @@ protected String getRequestMethod() {
+        archiveAction = getAction(courseArchiveRequest, submissionParams);
         result = getJsonResult(archiveAction);
 
         assertEquals(HttpStatus.SC_OK, result.getStatusCode());
```
Verify the instructor is still in archive status?

xpdavid commented on src/test/java/teammates/test/cases/webapi/ArchiveCourseActionTest.java:
```diff
@@ -28,71 +30,67 @@ protected String getRequestMethod() {
 
         assertEquals(HttpStatus.SC_OK, result.getStatusCode());
 
         ______TS("Typical case: unarchive a course, redirect to Courses page");
```
`redirect to Courses page`???

ChooJeremy replied:
Artifact of copying over the previous action test  🤣

xpdavid commented on src/test/java/teammates/test/cases/webapi/ArchiveCourseActionTest.java:
```diff
@@ -28,71 +30,67 @@ protected String getRequestMethod() {
         assertEquals(HttpStatus.SC_OK, result.getStatusCode());
+        assertFalse(theInstructor.isArchived);
 
         ______TS("Rare case: unarchive an active course, redirect to Courses page");
```
`redirect to Courses page`???

xpdavid commented on src/test/java/teammates/test/cases/webapi/ArchiveCourseActionTest.java:
```diff
@@ -28,71 +30,67 @@ protected String getRequestMethod() {
+        unarchiveAction = getAction(courseArchiveRequest, submissionParams);
         result = getJsonResult(unarchiveAction);
 
         assertEquals(HttpStatus.SC_OK, result.getStatusCode());
```
Verify the instructor is not in archive status?

xpdavid commented on src/test/java/teammates/test/cases/webapi/ArchiveCourseActionTest.java:
```diff
@@ -28,71 +30,67 @@ protected String getRequestMethod() {
 
         assertEquals(HttpStatus.SC_OK, result.getStatusCode());
 
         ______TS("Masquerade mode: archive course, redirect to Courses page");
```
`redirect to Courses page`???

xpdavid commented on src/test/java/teammates/test/cases/webapi/ArchiveCourseActionTest.java:
```diff
@@ -28,71 +30,67 @@ protected String getRequestMethod() {
+        archiveAction = getAction(courseArchiveRequest, addUserIdToParams(instructorId, submissionParams));
         result = getJsonResult(archiveAction);
 
         assertEquals(HttpStatus.SC_OK, result.getStatusCode());
```
Verify the instructor is in archive status?

xpdavid commented on src/test/java/teammates/test/cases/webapi/BinCourseActionTest.java:
```diff
@@ -0,0 +1,116 @@
+
+        verifyHttpParameterFailure();
+
+        ______TS("Typical case, 2 courses");
```
`2 courses`??

ChooJeremy replied:
It looks like this means: 
2 courses associated with instructor, bin 1 of them.
We expect that the course that was not binned is still in the course list.

xpdavid replied:
It is unlikely that binning one course will cause another course to be binned. It's ok to keep the test here. Just need to make the description clear and **add** a direct verification.

xpdavid commented on src/test/java/teammates/test/cases/webapi/BinCourseActionTest.java:
```diff
@@ -0,0 +1,116 @@
+                Const.ParamsNames.COURSE_ID, instructor1OfCourse1.courseId,
+        };
+
+        BinCourseAction deleteAction = getAction(submissionParams);
```
It is not `deleteAction`?

xpdavid commented on src/test/java/teammates/test/cases/webapi/BinCourseActionTest.java:
```diff
@@ -0,0 +1,116 @@
+
+        List<CourseAttributes> courseList = logic.getCoursesForInstructor(instructorId);
+        assertEquals(1, courseList.size());
+        assertEquals("icdct.tpa.id1", courseList.get(0).getId());
```
You should verify the course is marked as deleted?

Use `logic.getCourse` to retrieve the course you just "deleted"?

xpdavid commented on src/test/java/teammates/test/cases/webapi/BinCourseActionTest.java:
```diff
@@ -0,0 +1,116 @@
+        assertEquals("The course icdct.tpa.id1 has been deleted. You can restore it from the Recycle Bin manually.",
+                msg.getMessage());
+
+        courseList = logic.getCoursesForInstructor(instructorId);
```
The verification is implicit. Why not using `logic.getCourse` to verify the status directly?

xpdavid commented on src/test/java/teammates/test/cases/webapi/DeleteCourseActionTest.java:
```diff
@@ -30,87 +27,65 @@ protected String getRequestMethod() {
-        CoursesLogic.inst().createCourseAndInstructor(instructorId, "icdct.tpa.id1", "New course", "UTC");
         String[] submissionParams = new String[] {
-                Const.ParamsNames.COURSE_ID, instructor1OfCourse1.courseId,
+                Const.ParamsNames.INSTRUCTOR_ID, instructorId,
```
You don't need `INSTRUCTOR_ID` anymore right?

xpdavid commented on src/test/java/teammates/test/cases/webapi/DeleteCourseActionTest.java:
```diff
@@ -30,87 +27,65 @@ protected String getRequestMethod() {
-        submissionParams = new String[] {
-                Const.ParamsNames.COURSE_ID, "icdct.tpa.id1",
+        String[] submissionParams = new String[] {
+                Const.ParamsNames.INSTRUCTOR_ID, instructorId,
```
Same here?

xpdavid commented on src/test/java/teammates/test/cases/webapi/DeleteCourseActionTest.java:
```diff
@@ -30,87 +27,65 @@ protected String getRequestMethod() {
 
-        assertEquals(HttpStatus.SC_OK, result.getStatusCode());
+        DeleteCourseAction action = getAction(submissionParams);
+        action.checkSpecificAccessControl();
```
You should not call this method here in `testExecute()` test case.

xpdavid commented on src/test/java/teammates/test/cases/webapi/DeleteCourseActionTest.java:
```diff
@@ -30,87 +27,65 @@ protected String getRequestMethod() {
+        String courseId = instructor1OfCourse1.courseId;
 
-        loginAsAdmin();
+        ______TS("Error case, delete a course not in Recycle Bin");
```
It is not an error case?

xpdavid commented on src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.ts:
```diff
@@ -613,6 +608,89 @@ export class InstructorCourseEditPageComponent implements OnInit {
       editedInstructor.privileges.sessionLevel = newSessionLevelPrivileges;
     }
 
+    if (instr.controls.role.value === 'Custom') {
```
Are these change intended?

ChooJeremy replied:
The changes are made here: https://github.com/TEAMMATES/teammates/pull/9437/commits/4111125b924f147bb698bc44ff525f2ff4606c3a Not sure why it is showing it as a change...

xpdavid replied:
You need to remove irrelevant changes.

ChooJeremy replied:
I'll rebase my changes onto master to remove them then.

ChooJeremy commented:
Ready for review.

xpdavid commented on src/test/java/teammates/test/cases/webapi/DeleteCourseActionTest.java:
```diff
@@ -30,87 +27,62 @@ protected String getRequestMethod() {
-        assertEquals("The course idOfTypicalCourse1 has been deleted. You can restore it from the Recycle Bin manually.",
-                msg.getMessage());
+        assertEquals(HttpStatus.SC_OK, result.getStatusCode());
+        assertEquals("The course " + courseId + " has been permanently deleted.", message.getMessage());
```
Verify that the course is deleted. Similar in other places.

xpdavid commented on src/test/java/teammates/test/cases/webapi/SaveCourseActionTest.java:
```diff
@@ -74,13 +76,11 @@ public void testExecute() throws Exception {
+        courseEditSaveAction = getAction(courseSaveRequest, submissionParams);
         r = getJsonResult(courseEditSaveAction);
 
         assertEquals(HttpStatus.SC_OK, r.getStatusCode());
```
Verify the course is updated? Similar in other places.

ChooJeremy commented on src/test/java/teammates/test/cases/webapi/CreateCourseActionTest.java:
```diff
@@ -33,35 +34,30 @@ public void testExecute() throws Exception {
-        action = getAction(submissionParams);
+        action = getAction(courseCreateRequest);
         result = getJsonResult(action);
         MessageOutput message = (MessageOutput) result.getOutput();
```
You mean above, right? In this case the course isn't created but instead a course with the same ID as an existing course is attempted to be created. Unless you didn't make a mistake and git force pushing messed it up.

xpdavid replied:
> instead a course with the same ID as an existing course is attempted to be created

If you choose the courseId carefully, this will not happen.

> Unless you didn't make a mistake and git force pushing messed it up.

How does this relate to `git push --force`?

ChooJeremy commented:
Ready for review.

xpdavid commented on src/test/java/teammates/test/cases/webapi/DeleteCourseActionTest.java:
```diff
@@ -30,87 +27,66 @@ protected String getRequestMethod() {
+    public void testExecute_notInRecycleBin_shouldPass() throws Exception {
+        InstructorAttributes instructor1OfCourse1 = typicalBundle.instructors.get("instructor1OfCourse1");
+        String instructorId = instructor1OfCourse1.googleId;
+        String courseId = instructor1OfCourse1.courseId;
```
You need have assertion to make sure the course is not in recycle bin.

xpdavid commented on src/test/java/teammates/test/cases/webapi/DeleteCourseActionTest.java:
```diff
@@ -30,87 +27,66 @@ protected String getRequestMethod() {
-        assertEquals("icdct.tpa.id1", courseList.get(0).getId());
+        assertEquals(HttpStatus.SC_OK, result.getStatusCode());
+        assertEquals("The course " + courseId + " has been permanently deleted.", message.getMessage());
+        assertThrows(NullPointerException.class, () -> logic.getSoftDeletedCourseForInstructor(instructor1OfCourse1));
```
Rather than catch NPE, you can get the course directly using `logic.getCourse()` ?

ChooJeremy replied:
Sorry, how would this work? `logic.getCourse(instructor1OfCourse1.courseId)` returns null, do I just check that it returns null?

xpdavid replied:
Exactly. assertNull()

xpdavid commented on src/main/java/teammates/ui/webapi/request/CourseUpdateRequest.java:
```diff
@@ -0,0 +1,7 @@
+package teammates.ui.webapi.request;
+
+/**
+ * The save request for a course.
```
`The update request for a course.`

xpdavid commented on src/main/java/teammates/ui/webapi/output/CourseData.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * The API output format of a course.
+ */
+public class CourseData extends ApiOutput {
```
It looks like you haven't modify the frontend to use the Course type?

ChooJeremy replied:
Just to check with you, I'm thinking the following for API output, would they be ok?
DeleteCourse: return Course (only ID is used in frontend)
CreateCourse: return Course
UpdateCourse: return Course
ArchiveCourse: return a new CourseArchiveData object, which contains isArchived, the only variable used in frontend.
SoftDeleteCourse: return Course (only ID is used in frontend)

xpdavid replied:
Looks good. `CourseArchiveData` should contain `courseId` and `isArchived`?

xpdavid commented on src/main/java/teammates/ui/webapi/output/CoursesData.java:
```diff
@@ -0,0 +1,21 @@
+/**
+ * The API output format of a list of {@link CourseAttributes}.
+ */
+public class CoursesData extends ApiOutput {
```
Same here.

ChooJeremy replied:
This appears to be used in GET courses, which is outside the scope of this PR?

xpdavid replied:
Notice that. You can change it in your phase 2 PR.

ChooJeremy commented:
Ready for review.

I have tested the course parts of the instructor home page, instructor course page, course edit page including create, update, bin, archive and delete on the front end and they all work. Testing was done against master branch to ensure that functionality remained the same.

xpdavid commented on src/main/java/teammates/ui/webapi/action/BinCourseAction.java:
```diff
@@ -0,0 +1,45 @@
+
+            return new JsonResult(new CourseData(courseAttributes));
+        } catch (EntityDoesNotExistException e) {
+            return new JsonResult(e.getMessage(), HttpStatus.SC_INTERNAL_SERVER_ERROR);
```
Rethrow the exception as `EntityNotFoundException()`?

xpdavid commented on src/main/java/teammates/ui/webapi/action/DeleteCourseAction.java:
```diff
@@ -21,23 +20,20 @@ public void checkSpecificAccessControl() {
+
+        logic.deleteCourse(idOfCourseToDelete);
 
+        return new JsonResult(new CourseData(courseAttributes));
```
DELETE method should just a message right?

ChooJeremy replied:
A MessageOutput? I thought all API calls should be changed to not concern itself with what to be shown to the user.

xpdavid commented on src/main/java/teammates/ui/webapi/action/UpdateCourseAction.java:
```diff
@@ -58,7 +62,6 @@ public ActionResult execute() {
 
-        return new JsonResult("Updated course [" + courseId + "] details: Name: " + courseName
-                + ", Time zone: " + courseTimeZone, HttpStatus.SC_OK);
+        return new JsonResult(new CourseData(logic.getCourse(courseId)));
```
`updateCourseCascade` method will return the updated entity. There is no need to query it again here.

xpdavid commented on src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:
```diff
@@ -182,60 +184,46 @@ export class InstructorCoursesPageComponent implements OnInit {
+    }).subscribe((course: Course) => {
       this.loadInstructorCourses();
-      this.statusMessageService.showSuccessMessage(resp.message);
+      this.statusMessageService.showSuccessMessage('The course has been added. '
```
I don't think the HTML will work. I believe the previous implementation also does not work. You can leave it here. If you know how to fix it. Go ahead.

ChooJeremy replied:
It doesn't? It seems to work just fine for me.
![image](https://user-images.githubusercontent.com/4904988/53417149-e86eb280-3a0f-11e9-95bb-65873320b249.png)
The link brings you correctly to the course enroll page and the edit course details page. 

Or is this not intended behavior?

xpdavid replied:
The HTML code is not being escaped 😕

wkurniawan07 replied:
*was. The status message component has been extended to allow such case.

xpdavid commented on src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:
```diff
@@ -182,60 +184,46 @@ export class InstructorCoursesPageComponent implements OnInit {
 
   /**
-   * Unarchives an archived course.
+   * Archives an active course.
```
The doc does not match?

xpdavid commented on src/web/services/course.service.ts:
```diff
@@ -0,0 +1,57 @@
+  /**
+   * Archives a course by calling API.
+   */
+  archiveCourse(courseid: string, request: CourseArchiveRequest): Observable<CourseArchive> {
```
Confused name? `archiveCourse`. Just split this method into two? `archiveCourse` and `unarchiveCourse`? Or `changeArchiveStatus`?

ChooJeremy commented:
Ready for review.

wkurniawan07 commented:
@ChooJeremy you need to resolve the merge conflicts.

xpdavid commented on src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:
```diff
@@ -272,17 +257,16 @@ export class InstructorCoursesPageComponent implements OnInit {
-          });
+      this.courseService.deleteCourse(courseId).subscribe((message: MessageOutput) => {
+        this.loadInstructorCourses();
+        if (message.message === 'OK') {
```
This check is redundant as all non-2xx response will be redirect to error callback.

ChooJeremy replied:
Sure, and I've made the necessary changes. However, couldn't there be a case for having this check, as it is safer? In case the API is ever updated without the front-end changing, then if the API ever reports a value other than OK it may not be something the frontend just wants to process as a success.

wkurniawan07 replied:
@ChooJeremy there is an example [here](https://github.com/TEAMMATES/teammates/blob/2429ab644d751ff8e2fc310cd6dc854f40992868/src/main/java/teammates/ui/webapi/action/ConfirmFeedbackSessionSubmissionAction.java#L96-L104) for an action with different forms of success. There is clearly only one success case for the action being discussed here, so having an extra "OK" check does not value-add.

ChooJeremy commented:
Ready for review.

wkurniawan07 commented on src/main/java/teammates/ui/webapi/request/CourseCreateRequest.java:
```diff
@@ -0,0 +1,22 @@
+    @Override
+    public void validate() {
+        super.validate();
+        assertTrue(courseId != null, "Time zone should not be null");
```
This is not `time zone`?
